src/manager.S

```diff
@@ -0,0 +1,165 @@
+.data
+
+comps_cnt:     .dword 0
+comps_addr:    .dword 0
+switcher_caps: .dword 0
+
+.text
+.balign 4
+
+#include "comps_offsets.h"
+
+// Variables
+.global comps_cnt
+.global comps_addr
+.global switcher_caps
+
+// Functions
+.global asm_call_wrapper
+.global init_compartments
+.global add_compartment
+.global del_compartment
+
+/**
+ * Wrapper to store/restore state when coming from C
+ *
+ * @param x0 ASM function to call
+ * @param x1-x6 parameters to pass to function in c0
+ */
+.type asm_call_wrapper, "function"
+asm_call_wrapper:
+    // Save `x0` so we can temporarily use it
+    cvtp      c0, x0
+    str       c0, [sp]
+
+    // Derive `clr` (in case asm function does something weird with `PCC`)
+    cvtp      c0, lr
+    swp       c0, c0, [sp]
+    sub       sp, sp, #16
+
+    blr       c0
+    ldr       clr, [sp, #16]!
+    ret       clr
+
+/**
+ * Sets up memory for compartments
+ *
+ * Initializes required memory for compartments. This involves allocating a
+ * memory region for use by switcher code, to contain required capabilities,
+ * and deriving appropriate PCC/DDC values containing the executable switcher
+ * code, and the aforementioned memory region, respectively.
+ *
+ * @return Pointer to newly allocated memory region
+ */
+.type init_compartments, "function"
+init_compartments:
+    // Compute size of required memory, equivalent to `length` parameter of
+    // `mmap`
+    mov       x0, #COMP_SIZE
+    mov       x1, #MAX_COMP_COUNT
+    mov       x2, #32    // size of the 2 switcher capabilities
+    madd      x1, x0, x1, x2
+
+    // Store length and `lr` on stack, as we'll need them later
+    stp       x1, lr, [sp, #-16]!
+
+    // Allocate memory for switcher
+    mov       x0, xzr    // address
+                         // length - already stored in `x1`
+    mov       w2, #3     // prot == PROT_READ | PROT_WRITE
+    mov       w3, #4098  // flags == MAP_PRIVATE | MAP_ANONYMOUS
+    mov       w4, #-1    // fd
+    mov       w5, wzr    // offset
+    bl        mmap
+
+    // Restore length and `lr`
+    ldp       x1, lr, [sp], #16
+
+    // Save pointer to new allocated memory in `switcher_caps`
+    adr       x2, switcher_caps
+    str       x0, [x2]
+
+    // Derive DDC for switcher
+    cvtp      c2, x0
+    scbnds    c2, c2, x1
+
+    // Derive PCC for `switch_compartments` and friends
+    adr       x3, switcher_entry
+    adr       x4, switch_compartment_end
+    sub       x4, x4, x3
+    cvtp      c3, x3
+    scbndse   c3, c3, x4
+
+    // Store (DDC, PCC) at `switcher_caps`
+    ldr       x1, switcher_caps
+    stp       c2, c3, [x1], #32
```

Comment:
This is odd. From the comment, I'd expect `adr x1, switcher_caps`, but actually I think this needs another level of indirection. Perhaps `switcher_caps[0,1]` would be clearer.

Note that the value loaded here into `x1` should already be in `x0` (from the `mmap` call), so this load can be elided.

The `, #32` is clever and compact!

Reply:
So `adr x1, switcher_caps` would get the address where the `switcher_caps` variable is stored. What I want is the value stored at that address. I think the comment is slightly unclear. You are correct that we have it in `x0`. I've updated this in https://github.com/capablevms/cheri-morello-compartmentalisation/pull/1/commits/67ff6da6fe76bebab07d93ff26732ad1e188ee8d.